storage/innobase/handler/ha_innodb.cc

```diff
@@ -9013,7 +9012,7 @@ ha_innobase::index_read(
 		build_template(false);
 	}
 
-	if (key_ptr != NULL) {
+	if (key_len) {
```

Comment:
This statement is guarding a call to `row_sel_convert_mysql_key_to_innobase()`. Near the start of that function there is an assignment
```c++
	key_end = key_ptr + key_len;
```
In `clang -fsanitize=undefined` (`cmake -DWITH_UBSAN=ON`) builds this would lead to
```
runtime error: applying non-zero offset _ to null pointer
```
However, in GCC (up to 14.0.1) there is no check like this, and even after [MDEV-26272](https://jira.mariadb.org/browse/MDEV-26272) was fixed, there still are many bugs outside InnoDB that prevent us from using `clang -fsanitize=undefined`. There is a check that enforces `__attribute__((nonnull))` in both GCC and clang:
``` 
runtime error: null pointer passed as argument _, which is declared to never be null
```
I would address this concern by adding a missing `__attribute__((nonnull))` to the declaration of `row_sel_convert_mysql_key_to_innobase()`. All parameters are supposed to be nonnull.

Reply:
The calls in `ha_innobase::records_in_range()` do look like undefined behaviour to me, so I am fixing it. I will try to get the regression tests to pass with the added `__attribute__((nonnull))`.

Reply:
The conversion buffer `buf` can be passed as `nullptr`. a6b365af342969507314804ad7c65fdd78afc481 declares the other parameters as `nonnull`.